app/services/update_ocfl_service.rb

```diff
@@ -0,0 +1,94 @@
+class UpdateOcflService
+  class BlobError < StandardError; end
+  class RequestError < StandardError; end
+
+  def self.write!(...)
+    new(...).write!
+  end
+
+  def initialize(purl, file_uploads_map)
+    @purl = purl
+    @file_uploads_map = file_uploads_map
+  end
+
+  def write!
+    check_files_in_structural
+    check_signed_ids
+    add_files
+    remove_files
+    builder.save
+  end
+
+  private
+
+  attr_reader :purl, :file_uploads_map
+
+  delegate :cocina_object, :stacks_druid_path, to: :purl
+
+  def check_files_in_structural
+    return if file_uploads_map.keys.all? { |filename| cocina_filenames.include?(filename) }
+
+    raise RequestError, 'Files in file_uploads not in cocina object'
+  end
+
+  def check_signed_ids
+    return if file_uploads_map.values.all? { |signed_id| ActiveStorage.verifier.valid_message?(signed_id) }
+
+    raise RequestError, "Invalid signed ids found"
+  end
+
+  # Copy the files from ActiveStorage to ocfl
+  def add_files
+    file_uploads_map.each do |filename, signed_id|
+      blob = blob_for_signed_id(signed_id, filename)
+      blob_path = ActiveStorage::Blob.service.path_for(blob.key)
+
+      dir_name = File.dirname(filename)
+      builder.copy_file(blob_path, destination_path: dir_name == '.' ? nil : dir_name)
+    end
+  end
+
+  # return [ActiveStorage::Blob] the blob for the signed id
+  def blob_for_signed_id(signed_id, filename)
+    file_id = ActiveStorage.verifier.verified(signed_id, purpose: :blob_id)
+    ActiveStorage::Blob.find(file_id)
+  rescue ActiveRecord::RecordNotFound
+    raise BlobError, "Unable to find upload for #{filename} (#{signed_id})"
+  end
+
+  # Remove files from the ocfl that are not in the cocina object
+  def remove_files
+    ocfl_files_to_delete = directory.versions[directory.head].file_names.reject { |ocfl_file| cocina_filenames.include?(ocfl_file) }
```

Comment:
could we look at `builder.version.state.values.flatten` instead of `directory.versions[directory.head].file_names` ?


I don't like that we have both a handle on `builder` and `directory`